From patchwork Mon Jun 12 15:45:14 2017
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: [pulseaudio-discuss,
 1/3] alsa-mixer: add query_hw_device option to mappings
From: Tanu Kaskinen <tanuk@iki.fi>
X-Patchwork-Id: 161214
Message-Id: <20170612154516.12941-2-tanuk@iki.fi>
To: pulseaudio-discuss@lists.freedesktop.org
Cc: Takashi Iwai <tiwai@suse.de>, Hans de Goede <hdegoede@redhat.com>
Date: Mon, 12 Jun 2017 18:45:14 +0300

We have so far assumed that HDMI always uses device indexes 3, 7, 8, 9,
10, 11, 12 and 13. These values are hardcoded in the path configuration.
The Intel HDMI LPE driver, however, uses a different device numbering
scheme.

This patch adds a new configuration option for mappings. The option
tells PulseAudio to query the hw device index using snd_pcm_info(). It's
disabled by default, because querying the hw device doesn't always work
(but it's expected work for all hdmi devices).

This patch doesn't yet use the queried device index, that's left for
the subsequent patches.

BugLink: https://bugs.freedesktop.org/show_bug.cgi?id=100488
---
 src/modules/alsa/alsa-mixer.c                    | 83 +++++++++++++++++++++---
 src/modules/alsa/alsa-mixer.h                    | 12 ++++
 src/modules/alsa/mixer/profile-sets/default.conf | 37 +++++++++++
 3 files changed, 124 insertions(+), 8 deletions(-)

diff --git a/src/modules/alsa/alsa-mixer.c b/src/modules/alsa/alsa-mixer.c
index f59cad394..8bd90a728 100644
--- a/src/modules/alsa/alsa-mixer.c
+++ b/src/modules/alsa/alsa-mixer.c
@@ -3504,6 +3504,7 @@ pa_alsa_mapping *pa_alsa_mapping_get(pa_alsa_profile_set *ps, const char *name)
     pa_sample_spec_init(&m->sample_spec);
     pa_channel_map_init(&m->channel_map);
     m->proplist = pa_proplist_new();
+    m->hw_device_index = -1;
 
     pa_hashmap_put(ps->mappings, m->name, m);
 
@@ -3641,6 +3642,31 @@ static int mapping_parse_exact_channels(pa_config_parser_state *state) {
     return 0;
 }
 
+static int mapping_parse_query_hw_device(pa_config_parser_state *state) {
+    pa_alsa_profile_set *ps;
+    pa_alsa_mapping *m;
+    int b;
+
+    pa_assert(state);
+
+    ps = state->userdata;
+
+    if (!(m = pa_alsa_mapping_get(ps, state->section))) {
+        pa_log("[%s:%u] Option '%s' not expected in section '%s'",
+               state->filename, state->lineno, state->lvalue, state->section);
+        return -1;
+    }
+
+    if ((b = pa_parse_boolean(state->rvalue)) < 0) {
+        pa_log("[%s:%u] %s has invalid value '%s'", state->filename, state->lineno, state->lvalue, state->section);
+        return -1;
+    }
+
+    m->query_hw_device = b;
+
+    return 0;
+}
+
 static int mapping_parse_element(pa_config_parser_state *state) {
     pa_alsa_profile_set *ps;
     pa_alsa_mapping *m;
@@ -4357,6 +4383,7 @@ pa_alsa_profile_set* pa_alsa_profile_set_new(const char *fname, const pa_channel
         { "element-output",         mapping_parse_element,        NULL, NULL },
         { "direction",              mapping_parse_direction,      NULL, NULL },
         { "exact-channels",         mapping_parse_exact_channels, NULL, NULL },
+        { "query-hw-device",        mapping_parse_query_hw_device,NULL, NULL },
 
         /* Shared by [Mapping ...] and [Profile ...] */
         { "description",            mapping_parse_description,    NULL, NULL },
@@ -4531,6 +4558,24 @@ static int add_profiles_to_probe(
     return i;
 }
 
+static int mapping_query_hw_device(pa_alsa_mapping *mapping, snd_pcm_t *pcm) {
+    int r;
+    snd_pcm_info_t* pcm_info;
+    snd_pcm_info_alloca(&pcm_info);
+
+    pa_assert(mapping->query_hw_device);
+
+    r = snd_pcm_info(pcm, pcm_info);
+    if (r < 0) {
+        pa_log("Mapping %s has query_hw_device enabled, but snd_pcm_info() failed %s: ", mapping->name, pa_alsa_strerror(r));
+        return -1;
+    }
+
+    mapping->hw_device_index = snd_pcm_info_get_device(pcm_info);
+
+    return 0;
+}
+
 void pa_alsa_profile_set_probe(
         pa_alsa_profile_set *ps,
         const char *dev_id,
@@ -4574,6 +4619,7 @@ void pa_alsa_profile_set_probe(
 
         /* Skip if this is already marked that it is supported (i.e. from the config file) */
         if (!p->supported) {
+            bool only_one_mapping;
 
             profile_finalize_probing(last, p);
             p->supported = true;
@@ -4601,6 +4647,15 @@ void pa_alsa_profile_set_probe(
             if (p->supported)
                 pa_log_debug("Looking at profile %s", p->name);
 
+            if (p->output_mappings && pa_idxset_size(p->output_mappings) == 1 &&
+                ((!p->input_mappings) || pa_idxset_size(p->input_mappings) == 0))
+                only_one_mapping = true;
+            else if (p->input_mappings && pa_idxset_size(p->input_mappings) == 1 &&
+                     ((!p->output_mappings) || pa_idxset_size(p->output_mappings) == 0))
+                only_one_mapping = true;
+            else
+                only_one_mapping = false;
+
             /* Check if we can open all new ones */
             if (p->output_mappings && p->supported)
                 PA_IDXSET_FOREACH(m, p->output_mappings, idx) {
@@ -4612,11 +4667,17 @@ void pa_alsa_profile_set_probe(
                     if (!(m->output_pcm = mapping_open_pcm(m, ss, dev_id, m->exact_channels,
                                                            SND_PCM_STREAM_PLAYBACK,
                                                            default_n_fragments,
-                                                           default_fragment_size_msec))) {
+                                                           default_fragment_size_msec)))
                         p->supported = false;
-                        if (pa_idxset_size(p->output_mappings) == 1 &&
-                            ((!p->input_mappings) || pa_idxset_size(p->input_mappings) == 0)) {
-                            pa_log_debug("Caching failure to open output:%s", m->name);
+
+                    if (m->output_pcm && m->query_hw_device && m->hw_device_index < 0) {
+                        if (mapping_query_hw_device(m, m->output_pcm) < 0)
+                            p->supported = false;
+                    }
+
+                    if (!p->supported) {
+                        if (only_one_mapping) {
+                            pa_log_debug("Caching failure to use output:%s", m->name);
                             pa_hashmap_put(broken_outputs, m, m);
                         }
                         break;
@@ -4633,11 +4694,17 @@ void pa_alsa_profile_set_probe(
                     if (!(m->input_pcm = mapping_open_pcm(m, ss, dev_id, m->exact_channels,
                                                           SND_PCM_STREAM_CAPTURE,
                                                           default_n_fragments,
-                                                          default_fragment_size_msec))) {
+                                                          default_fragment_size_msec)))
                         p->supported = false;
-                        if (pa_idxset_size(p->input_mappings) == 1 &&
-                            ((!p->output_mappings) || pa_idxset_size(p->output_mappings) == 0)) {
-                            pa_log_debug("Caching failure to open input:%s", m->name);
+
+                    if (m->input_pcm && m->query_hw_device && m->hw_device_index < 0) {
+                        if (mapping_query_hw_device(m, m->input_pcm) < 0)
+                            p->supported = false;
+                    }
+
+                    if (!p->supported) {
+                        if (only_one_mapping) {
+                            pa_log_debug("Caching failure to use input:%s", m->name);
                             pa_hashmap_put(broken_inputs, m, m);
                         }
                         break;
diff --git a/src/modules/alsa/alsa-mixer.h b/src/modules/alsa/alsa-mixer.h
index 4ebf1922b..96678aea4 100644
--- a/src/modules/alsa/alsa-mixer.h
+++ b/src/modules/alsa/alsa-mixer.h
@@ -275,6 +275,18 @@ struct pa_alsa_mapping {
     bool exact_channels:1;
     bool fallback:1;
 
+    /* Should we try to figure out which hw device this mapping is using? HDMI
+     * jack detection and ELD handling requires us to know the underlying hw
+     * device. This defaults to false, because we only need this for HDMI, and
+     * it's not always possible to figure out the hw device (for example, a
+     * surround device might consist of several stereo hw devices). */
+    bool query_hw_device;
+
+    /* The "y" in "hw:x,y". This is set to -1 before the device index has been
+     * queried, or if the query failed. If query_hw_device is false, this is
+     * always -1. */
+    int hw_device_index;
+
     /* Temporarily used during probing */
     snd_pcm_t *input_pcm;
     snd_pcm_t *output_pcm;
diff --git a/src/modules/alsa/mixer/profile-sets/default.conf b/src/modules/alsa/mixer/profile-sets/default.conf
index f412058ff..01203b3c3 100644
--- a/src/modules/alsa/mixer/profile-sets/default.conf
+++ b/src/modules/alsa/mixer/profile-sets/default.conf
@@ -57,6 +57,10 @@
 ; exact-channels = yes | no                 # If no, and the exact number of channels is not supported,
 ;                                           # allow device to be opened with another channel count
 ; fallback = no | yes                       # This mapping will only be considered if all non-fallback mappings fail
+; query-hw-device = no | yes                # Defaults to "no", needs to be set to "yes" for HDMI mappings. HDMI
+;                                           # jack detection and ELD handling requires us to know which hw:x,y
+;                                           # device the hdmi:x,z device corresponds to. This defaults to "no",
+;                                           # because querying the hw device doesn't work for all devices.
 ; [Profile id]
 ; input-mappings = ...                      # Lists mappings for sources on this profile, those mapping must be
 ;                                           # defined in this file too
@@ -164,6 +168,7 @@ channel-map = left,right
 paths-input = iec958-stereo-input
 paths-output = iec958-stereo-output
 priority = 5
+query-hw-device = yes
 
 [Mapping iec958-ac3-surround-40]
 device-strings = a52:%f
@@ -193,6 +198,7 @@ paths-output = hdmi-output-0
 channel-map = left,right
 priority = 4
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround]
 description = Digital Surround 5.1 (HDMI)
@@ -201,6 +207,7 @@ paths-output = hdmi-output-0
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 3
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71]
 description = Digital Surround 7.1 (HDMI)
@@ -209,6 +216,7 @@ paths-output = hdmi-output-0
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 3
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround]
 description = Digital Surround 5.1 (HDMI/DTS)
@@ -217,6 +225,7 @@ paths-output = hdmi-output-0
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra1]
 description = Digital Stereo (HDMI 2)
@@ -225,6 +234,7 @@ paths-output = hdmi-output-1
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra1]
 description = Digital Surround 5.1 (HDMI 2)
@@ -233,6 +243,7 @@ paths-output = hdmi-output-1
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra1]
 description = Digital Surround 7.1 (HDMI 2)
@@ -241,6 +252,7 @@ paths-output = hdmi-output-1
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra1]
 description = Digital Surround 5.1 (HDMI 2/DTS)
@@ -249,6 +261,7 @@ paths-output = hdmi-output-1
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra2]
 description = Digital Stereo (HDMI 3)
@@ -257,6 +270,7 @@ paths-output = hdmi-output-2
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra2]
 description = Digital Surround 5.1 (HDMI 3)
@@ -265,6 +279,7 @@ paths-output = hdmi-output-2
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra2]
 description = Digital Surround 7.1 (HDMI 3)
@@ -273,6 +288,7 @@ paths-output = hdmi-output-2
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra2]
 description = Digital Surround 5.1 (HDMI 3/DTS)
@@ -281,6 +297,7 @@ paths-output = hdmi-output-2
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra3]
 description = Digital Stereo (HDMI 4)
@@ -289,6 +306,7 @@ paths-output = hdmi-output-3
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra3]
 description = Digital Surround 5.1 (HDMI 4)
@@ -297,6 +315,7 @@ paths-output = hdmi-output-3
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra3]
 description = Digital Surround 7.1 (HDMI 4)
@@ -305,6 +324,7 @@ paths-output = hdmi-output-3
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra3]
 description = Digital Surround 5.1 (HDMI 4/DTS)
@@ -313,6 +333,7 @@ paths-output = hdmi-output-3
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra4]
 description = Digital Stereo (HDMI 5)
@@ -321,6 +342,7 @@ paths-output = hdmi-output-4
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra4]
 description = Digital Surround 5.1 (HDMI 5)
@@ -329,6 +351,7 @@ paths-output = hdmi-output-4
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra4]
 description = Digital Surround 7.1 (HDMI 5)
@@ -337,6 +360,7 @@ paths-output = hdmi-output-4
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra4]
 description = Digital Surround 5.1 (HDMI 5/DTS)
@@ -345,6 +369,7 @@ paths-output = hdmi-output-4
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra5]
 description = Digital Stereo (HDMI 6)
@@ -353,6 +378,7 @@ paths-output = hdmi-output-5
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra5]
 description = Digital Surround 5.1 (HDMI 6)
@@ -361,6 +387,7 @@ paths-output = hdmi-output-5
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra5]
 description = Digital Surround 7.1 (HDMI 6)
@@ -369,6 +396,7 @@ paths-output = hdmi-output-5
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra5]
 description = Digital Surround 5.1 (HDMI 6/DTS)
@@ -377,6 +405,7 @@ paths-output = hdmi-output-5
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra6]
 description = Digital Stereo (HDMI 7)
@@ -385,6 +414,7 @@ paths-output = hdmi-output-6
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra6]
 description = Digital Surround 5.1 (HDMI 7)
@@ -393,6 +423,7 @@ paths-output = hdmi-output-6
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra6]
 description = Digital Surround 7.1 (HDMI 7)
@@ -401,6 +432,7 @@ paths-output = hdmi-output-6
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra6]
 description = Digital Surround 5.1 (HDMI 7/DTS)
@@ -409,6 +441,7 @@ paths-output = hdmi-output-6
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-stereo-extra7]
 description = Digital Stereo (HDMI 8)
@@ -417,6 +450,7 @@ paths-output = hdmi-output-7
 channel-map = left,right
 priority = 2
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround-extra7]
 description = Digital Surround 5.1 (HDMI 8)
@@ -425,6 +459,7 @@ paths-output = hdmi-output-7
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-surround71-extra7]
 description = Digital Surround 7.1 (HDMI 8)
@@ -433,6 +468,7 @@ paths-output = hdmi-output-7
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe,side-left,side-right
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping hdmi-dts-surround-extra7]
 description = Digital Surround 5.1 (HDMI 8/DTS)
@@ -441,6 +477,7 @@ paths-output = hdmi-output-7
 channel-map = front-left,front-right,rear-left,rear-right,front-center,lfe
 priority = 1
 direction = output
+query-hw-device = yes
 
 [Mapping multichannel-output]
 device-strings = hw:%f

From patchwork Mon Jun 12 15:45:15 2017
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: [pulseaudio-discuss,
 2/3] alsa-mixer: implement ELD device autodetection
From: Tanu Kaskinen <tanuk@iki.fi>
X-Patchwork-Id: 161216
Message-Id: <20170612154516.12941-3-tanuk@iki.fi>
To: pulseaudio-discuss@lists.freedesktop.org
Cc: Takashi Iwai <tiwai@suse.de>, Hans de Goede <hdegoede@redhat.com>
Date: Mon, 12 Jun 2017 18:45:15 +0300

This removes the need to hardcode the ELD device index in the path
configuration. The hardcoded values don't work with the Intel HDMI LPE
driver.

BugLink: https://bugs.freedesktop.org/show_bug.cgi?id=100488
---
 src/modules/alsa/alsa-mixer.c                      | 31 +++++++++++++++++++---
 src/modules/alsa/alsa-mixer.h                      |  1 +
 .../alsa/mixer/paths/analog-output.conf.common     |  6 +++--
 src/modules/alsa/mixer/paths/hdmi-output-0.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-1.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-2.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-3.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-4.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-5.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-6.conf    |  2 +-
 src/modules/alsa/mixer/paths/hdmi-output-7.conf    |  2 +-
 11 files changed, 40 insertions(+), 14 deletions(-)

diff --git a/src/modules/alsa/alsa-mixer.c b/src/modules/alsa/alsa-mixer.c
index 8bd90a728..71b393e49 100644
--- a/src/modules/alsa/alsa-mixer.c
+++ b/src/modules/alsa/alsa-mixer.c
@@ -2032,6 +2032,28 @@ static int element_parse_enumeration(pa_config_parser_state *state) {
     return 0;
 }
 
+static int parse_eld_device(pa_config_parser_state *state) {
+    pa_alsa_path *path;
+    uint32_t eld_device;
+
+    path = state->userdata;
+
+    if (pa_atou(state->rvalue, &eld_device) >= 0) {
+        path->autodetect_eld_device = false;
+        path->eld_device = eld_device;
+        return 0;
+    }
+
+    if (pa_streq(state->rvalue, "auto")) {
+        path->autodetect_eld_device = true;
+        path->eld_device = -1;
+        return 0;
+    }
+
+    pa_log("[%s:%u] Invalid value for option 'eld-device': %s", state->filename, state->lineno, state->rvalue);
+    return -1;
+}
+
 static int option_parse_priority(pa_config_parser_state *state) {
     pa_alsa_path *p;
     pa_alsa_option *o;
@@ -2524,7 +2546,7 @@ pa_alsa_path* pa_alsa_path_new(const char *paths_dir, const char *fname, pa_alsa
         { "description-key",     pa_config_parse_string,            NULL, "General" },
         { "description",         pa_config_parse_string,            NULL, "General" },
         { "mute-during-activation", pa_config_parse_bool,           NULL, "General" },
-        { "eld-device",          pa_config_parse_int,               NULL, "General" },
+        { "eld-device",          parse_eld_device,                  NULL, "General" },
 
         /* [Option ...] */
         { "priority",            option_parse_priority,             NULL, NULL },
@@ -2563,7 +2585,6 @@ pa_alsa_path* pa_alsa_path_new(const char *paths_dir, const char *fname, pa_alsa
     items[1].data = &p->description_key;
     items[2].data = &p->description;
     items[3].data = &mute_during_activation;
-    items[4].data = &p->eld_device;
 
     if (!paths_dir)
         paths_dir = get_default_paths_dir();
@@ -3992,9 +4013,11 @@ static void mapping_paths_probe(pa_alsa_mapping *m, pa_alsa_profile *profile,
     }
 
     PA_HASHMAP_FOREACH(p, ps->paths, state) {
-        if (pa_alsa_path_probe(p, mixer_handle, m->profile_set->ignore_dB) < 0) {
+        if (p->autodetect_eld_device)
+            p->eld_device = m->hw_device_index;
+
+        if (pa_alsa_path_probe(p, m, mixer_handle, m->profile_set->ignore_dB) < 0)
             pa_hashmap_remove(ps->paths, p);
-        }
     }
 
     path_set_condense(ps, mixer_handle);
diff --git a/src/modules/alsa/alsa-mixer.h b/src/modules/alsa/alsa-mixer.h
index 96678aea4..d85cc38d3 100644
--- a/src/modules/alsa/alsa-mixer.h
+++ b/src/modules/alsa/alsa-mixer.h
@@ -191,6 +191,7 @@ struct pa_alsa_path {
     char *description_key;
     char *description;
     unsigned priority;
+    bool autodetect_eld_device;
     int eld_device;
     pa_proplist *proplist;
 
diff --git a/src/modules/alsa/mixer/paths/analog-output.conf.common b/src/modules/alsa/mixer/paths/analog-output.conf.common
index 17b45278a..a113a2e3a 100644
--- a/src/modules/alsa/mixer/paths/analog-output.conf.common
+++ b/src/modules/alsa/mixer/paths/analog-output.conf.common
@@ -64,8 +64,10 @@
 ; mute-during-activation = yes | no      # If this path supports hardware mute, should the hw mute be used while activating this
 ;                                        # path? In some cases this can reduce extra noises during port switching, while in other
 ;                                        # cases this can increase such noises. Default: no.
-; eld-device = ...                       # If this is an HDMI port, here's where to specify the device number for the ELD mixer
-;                                        # control. The default is to not make use of ELD information.
+; eld-device = ...                       # If this is an HDMI port, here's where to specify the device index for the ELD mixer
+;                                        # control. The default is to not make use of ELD information. If the value is "auto",
+;                                        # the device index is autodetected (requires setting "query-hw-device = yes" in the
+;                                        # mapping that contains this path).
 ;
 ; [Properties]                           # Property list for this path. The list is merged into the port property list.
 ; <key> = <value>                        # Each property is defined on its own line.
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-0.conf b/src/modules/alsa/mixer/paths/hdmi-output-0.conf
index 331014709..ce96a4989 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-0.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-0.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort
 priority = 59
-eld-device = 3
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-1.conf b/src/modules/alsa/mixer/paths/hdmi-output-1.conf
index d81ee789c..d35f6cb93 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-1.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-1.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 2
 priority = 58
-eld-device = 7
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-2.conf b/src/modules/alsa/mixer/paths/hdmi-output-2.conf
index 349812fc2..b54da82f4 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-2.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-2.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 3
 priority = 57
-eld-device = 8
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-3.conf b/src/modules/alsa/mixer/paths/hdmi-output-3.conf
index 81463c946..27f786fc7 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-3.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-3.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 4
 priority = 56
-eld-device = 9
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-4.conf b/src/modules/alsa/mixer/paths/hdmi-output-4.conf
index d61ec7547..8e7f82f40 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-4.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-4.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 5
 priority = 55
-eld-device = 10
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-5.conf b/src/modules/alsa/mixer/paths/hdmi-output-5.conf
index 02c15e893..708e83876 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-5.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-5.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 6
 priority = 54
-eld-device = 11
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-6.conf b/src/modules/alsa/mixer/paths/hdmi-output-6.conf
index 188a1adb3..340d5e495 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-6.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-6.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 7
 priority = 53
-eld-device = 12
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-7.conf b/src/modules/alsa/mixer/paths/hdmi-output-7.conf
index 80f4e3722..9946eda2c 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-7.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-7.conf
@@ -1,7 +1,7 @@
 [General]
 description = HDMI / DisplayPort 8
 priority = 52
-eld-device = 13
+eld-device = auto
 
 [Properties]
 device.icon_name = video-display

From patchwork Mon Jun 12 15:45:16 2017
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
Subject: [pulseaudio-discuss,
 3/3] alsa-mixer: autodetect the HDMI jack PCM device
From: Tanu Kaskinen <tanuk@iki.fi>
X-Patchwork-Id: 161215
Message-Id: <20170612154516.12941-4-tanuk@iki.fi>
To: pulseaudio-discuss@lists.freedesktop.org
Cc: Takashi Iwai <tiwai@suse.de>, Hans de Goede <hdegoede@redhat.com>
Date: Mon, 12 Jun 2017 18:45:16 +0300

This removes the need to hardcode the PCM device index in the HDMI jack
names. The hardcoded values don't work with the Intel HDMI LPE driver.

BugLink: https://bugs.freedesktop.org/show_bug.cgi?id=100488
---
 src/modules/alsa/alsa-mixer.c                      | 50 ++++++++++++++++++++--
 src/modules/alsa/alsa-mixer.h                      |  4 +-
 src/modules/alsa/alsa-sink.c                       |  2 +-
 src/modules/alsa/alsa-source.c                     |  2 +-
 .../alsa/mixer/paths/analog-output.conf.common     |  5 +++
 src/modules/alsa/mixer/paths/hdmi-output-0.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-1.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-2.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-3.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-4.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-5.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-6.conf    |  3 +-
 src/modules/alsa/mixer/paths/hdmi-output-7.conf    |  3 +-
 13 files changed, 73 insertions(+), 14 deletions(-)

diff --git a/src/modules/alsa/alsa-mixer.c b/src/modules/alsa/alsa-mixer.c
index 71b393e49..49ce39e6b 100644
--- a/src/modules/alsa/alsa-mixer.c
+++ b/src/modules/alsa/alsa-mixer.c
@@ -1812,12 +1812,31 @@ static int element_probe(pa_alsa_element *e, snd_mixer_t *m) {
     return 0;
 }
 
-static int jack_probe(pa_alsa_jack *j, snd_mixer_t *m) {
+static int jack_probe(pa_alsa_jack *j, pa_alsa_mapping *mapping, snd_mixer_t *m) {
     bool has_control;
 
     pa_assert(j);
     pa_assert(j->path);
 
+    if (j->append_pcm_to_name) {
+        char *new_name;
+
+        if (!mapping) {
+            /* This could also be an assertion, because this should never
+             * happen. At the time of writing, mapping can only be NULL when
+             * module-alsa-sink/source synthesizes a path, and those
+             * synthesized paths never have any jacks, so jack_probe() should
+             * never be called with a NULL mapping. */
+            pa_log("Jack %s: append_pcm_to_name is set, but mapping is NULL. Can't use this jack.", j->name);
+            return -1;
+        }
+
+        new_name = pa_sprintf_malloc("%s,pcm=%i Jack", j->name, mapping->hw_device_index);
+        pa_xfree(j->alsa_name);
+        j->alsa_name = new_name;
+        j->append_pcm_to_name = false;
+    }
+
     has_control = pa_alsa_mixer_find(m, j->alsa_name, 0) != NULL;
     pa_alsa_jack_set_has_control(j, has_control);
 
@@ -2348,6 +2367,30 @@ static int jack_parse_state(pa_config_parser_state *state) {
     return 0;
 }
 
+static int jack_parse_append_pcm_to_name(pa_config_parser_state *state) {
+    pa_alsa_path *path;
+    pa_alsa_jack *jack;
+    int b;
+
+    pa_assert(state);
+
+    path = state->userdata;
+    if (!(jack = jack_get(path, state->section))) {
+        pa_log("[%s:%u] Option 'append_pcm_to_name' not expected in section '%s'",
+               state->filename, state->lineno, state->section);
+        return -1;
+    }
+
+    b = pa_parse_boolean(state->rvalue);
+    if (b < 0) {
+        pa_log("[%s:%u] Invalid value for 'append_pcm_to_name': %s", state->filename, state->lineno, state->rvalue);
+        return -1;
+    }
+
+    jack->append_pcm_to_name = b;
+    return 0;
+}
+
 static int element_set_option(pa_alsa_element *e, snd_mixer_t *m, int alsa_idx) {
     snd_mixer_selem_id_t *sid;
     snd_mixer_elem_t *me;
@@ -2555,6 +2598,7 @@ pa_alsa_path* pa_alsa_path_new(const char *paths_dir, const char *fname, pa_alsa
         /* [Jack ...] */
         { "state.plugged",       jack_parse_state,                  NULL, NULL },
         { "state.unplugged",     jack_parse_state,                  NULL, NULL },
+        { "append-pcm-to-name",  jack_parse_append_pcm_to_name,     NULL, NULL },
 
         /* [Element ...] */
         { "switch",              element_parse_switch,              NULL, NULL },
@@ -2766,7 +2810,7 @@ static void path_create_settings(pa_alsa_path *p) {
     element_create_settings(p->elements, NULL);
 }
 
-int pa_alsa_path_probe(pa_alsa_path *p, snd_mixer_t *m, bool ignore_dB) {
+int pa_alsa_path_probe(pa_alsa_path *p, pa_alsa_mapping *mapping, snd_mixer_t *m, bool ignore_dB) {
     pa_alsa_element *e;
     pa_alsa_jack *j;
     double min_dB[PA_CHANNEL_POSITION_MAX], max_dB[PA_CHANNEL_POSITION_MAX];
@@ -2786,7 +2830,7 @@ int pa_alsa_path_probe(pa_alsa_path *p, snd_mixer_t *m, bool ignore_dB) {
     pa_log_debug("Probing path '%s'", p->name);
 
     PA_LLIST_FOREACH(j, p->jacks) {
-        if (jack_probe(j, m) < 0) {
+        if (jack_probe(j, mapping, m) < 0) {
             p->supported = false;
             pa_log_debug("Probe of jack '%s' failed.", j->alsa_name);
             return -1;
diff --git a/src/modules/alsa/alsa-mixer.h b/src/modules/alsa/alsa-mixer.h
index d85cc38d3..7400064fa 100644
--- a/src/modules/alsa/alsa-mixer.h
+++ b/src/modules/alsa/alsa-mixer.h
@@ -171,6 +171,8 @@ struct pa_alsa_jack {
 
     pa_dynarray *ucm_devices; /* pa_alsa_ucm_device */
     pa_dynarray *ucm_hw_mute_devices; /* pa_alsa_ucm_device */
+
+    bool append_pcm_to_name;
 };
 
 pa_alsa_jack *pa_alsa_jack_new(pa_alsa_path *path, const char *name);
@@ -235,7 +237,7 @@ void pa_alsa_element_dump(pa_alsa_element *e);
 
 pa_alsa_path *pa_alsa_path_new(const char *paths_dir, const char *fname, pa_alsa_direction_t direction);
 pa_alsa_path *pa_alsa_path_synthesize(const char *element, pa_alsa_direction_t direction);
-int pa_alsa_path_probe(pa_alsa_path *p, snd_mixer_t *m, bool ignore_dB);
+int pa_alsa_path_probe(pa_alsa_path *p, pa_alsa_mapping *mapping, snd_mixer_t *m, bool ignore_dB);
 void pa_alsa_path_dump(pa_alsa_path *p);
 int pa_alsa_path_get_volume(pa_alsa_path *p, snd_mixer_t *m, const pa_channel_map *cm, pa_cvolume *v);
 int pa_alsa_path_get_mute(pa_alsa_path *path, snd_mixer_t *m, bool *muted);
diff --git a/src/modules/alsa/alsa-sink.c b/src/modules/alsa/alsa-sink.c
index 827a65081..99ca5061b 100644
--- a/src/modules/alsa/alsa-sink.c
+++ b/src/modules/alsa/alsa-sink.c
@@ -1912,7 +1912,7 @@ static void find_mixer(struct userdata *u, pa_alsa_mapping *mapping, const char
         if (!(u->mixer_path = pa_alsa_path_synthesize(element, PA_ALSA_DIRECTION_OUTPUT)))
             goto fail;
 
-        if (pa_alsa_path_probe(u->mixer_path, u->mixer_handle, ignore_dB) < 0)
+        if (pa_alsa_path_probe(u->mixer_path, NULL, u->mixer_handle, ignore_dB) < 0)
             goto fail;
 
         pa_log_debug("Probed mixer path %s:", u->mixer_path->name);
diff --git a/src/modules/alsa/alsa-source.c b/src/modules/alsa/alsa-source.c
index 6bec188ea..84abbf1d9 100644
--- a/src/modules/alsa/alsa-source.c
+++ b/src/modules/alsa/alsa-source.c
@@ -1615,7 +1615,7 @@ static void find_mixer(struct userdata *u, pa_alsa_mapping *mapping, const char
         if (!(u->mixer_path = pa_alsa_path_synthesize(element, PA_ALSA_DIRECTION_INPUT)))
             goto fail;
 
-        if (pa_alsa_path_probe(u->mixer_path, u->mixer_handle, ignore_dB) < 0)
+        if (pa_alsa_path_probe(u->mixer_path, NULL, u->mixer_handle, ignore_dB) < 0)
             goto fail;
 
         pa_log_debug("Probed mixer path %s:", u->mixer_path->name);
diff --git a/src/modules/alsa/mixer/paths/analog-output.conf.common b/src/modules/alsa/mixer/paths/analog-output.conf.common
index a113a2e3a..13d4f7c3b 100644
--- a/src/modules/alsa/mixer/paths/analog-output.conf.common
+++ b/src/modules/alsa/mixer/paths/analog-output.conf.common
@@ -124,6 +124,11 @@
 ;                                      # the required-any are present.
 ; state.plugged = yes | no | unknown   # Normally a plugged jack would mean the port becomes available, and an unplugged means it's
 ; state.unplugged = yes | no | unknown # unavailable, but the port status can be overridden by specifying state.plugged and/or state.unplugged.
+; append-pcm-to-name = no | yes        # Add ",pcm=N" to the jack name? N is the hw PCM device index. HDMI jacks have
+;                                      # the PCM device index in their name, but different drivers use different
+;                                      # numbering schemes, so we can't hardcode the full jack name in our configuration
+;                                      # files. If this is set to "yes", then also "query-hw-device" in the mapping
+;                                      # configuration needs to be set to "yes".
 
 [Element PCM]
 switch = mute
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-0.conf b/src/modules/alsa/mixer/paths/hdmi-output-0.conf
index ce96a4989..95b1342e9 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-0.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-0.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=3]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-1.conf b/src/modules/alsa/mixer/paths/hdmi-output-1.conf
index d35f6cb93..37b945204 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-1.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-1.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=7]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-2.conf b/src/modules/alsa/mixer/paths/hdmi-output-2.conf
index b54da82f4..19c38f2e8 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-2.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-2.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=8]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-3.conf b/src/modules/alsa/mixer/paths/hdmi-output-3.conf
index 27f786fc7..8551570ac 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-3.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-3.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=9]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-4.conf b/src/modules/alsa/mixer/paths/hdmi-output-4.conf
index 8e7f82f40..e36128921 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-4.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-4.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=10]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-5.conf b/src/modules/alsa/mixer/paths/hdmi-output-5.conf
index 708e83876..82dc3be79 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-5.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-5.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=11]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-6.conf b/src/modules/alsa/mixer/paths/hdmi-output-6.conf
index 340d5e495..92e8fd1e2 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-6.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-6.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=12]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
diff --git a/src/modules/alsa/mixer/paths/hdmi-output-7.conf b/src/modules/alsa/mixer/paths/hdmi-output-7.conf
index 9946eda2c..abe2b60e6 100644
--- a/src/modules/alsa/mixer/paths/hdmi-output-7.conf
+++ b/src/modules/alsa/mixer/paths/hdmi-output-7.conf
@@ -6,5 +6,6 @@ eld-device = auto
 [Properties]
 device.icon_name = video-display
 
-[Jack HDMI/DP,pcm=13]
+[Jack HDMI/DP]
+append-pcm-to-name = yes
 required = ignore
